tests/python/pants_test/engine/test_rules.py

```diff
@@ -670,6 +674,43 @@ def b_from_d(d):
                      }""").strip(),
                                     subgraph)
 
+  def test_validate_yield_statements(self):
+    expected_rx_str = re.escape("""\
+Yield(value=Call(func=Name(id='A', ctx=Load()), args=[], keywords=[], starargs=None, kwargs=None))
+""")
+    with self.assertRaisesRegexp(_RuleVisitor.YieldVisitError, expected_rx_str):
+      @rule(A, [])
+      def f():
+        yield A()
+        # The yield statement isn't at the end.
+        return
+
+    with self.assertRaises(_RuleVisitor.YieldVisitError) as cm:
+      @rule(A, [])
+      def g():
+        # This is a yield statement without an assignment.
+        yield Get(B, D, D())
+        yield A()
+    exc_msg = str(cm.exception)
+    # Correctly matches the function name.
+    self.assertIn('In function g:', exc_msg)
+    # Matches the line number of the @rule decorator.
+    match_regexp = re.compile(r'^.*test_rules.py:{}:'.format(sys._getframe().f_lineno - 9),
+                              flags=re.MULTILINE)
+    assertRegex(self, exc_msg, match_regexp)
+    # Shows sufficient context around the rule definition to locate it.
+    self.assertIn("""\
+      @rule(A, [])
+      def g():
+""", exc_msg)
+    # Prints an AST dump of the offending yield statement for debuggability.
```

Comment:
I'm not sure I like the debug output being present in the exception message. If we can't render the relevant line verbatim, I think it would be preferable to just render the line number alone.

Reply:
The reason I did this is because I wasn't able to figure out how to get the line actually causing the error. I agree it's not good to have the debug output here. It's definitely possible to do this with the `ast` module, but it would require some minor bookkeeping (tracking line offsets from the function def itself). I'll look into how reasonable that is today.

Reply:
To be clear: I think it's plenty reasonable to just render the line number.